### PR TITLE
Speed up `unchop()`

### PR DESCRIPTION
**Pull request by mgirlich**
``` r
library(tidyr)
set.seed(1)

n <- 100e3
df <- tibble(
  a = purrr::map(1:n, ~ rlang::seq2(1, sample(0:10, 1))),
  y = 1:n
)

bench::mark(
  drop_empty = unchop(df, a),
  keep_empty = unchop(df, a, keep_empty = TRUE),
  check = FALSE
)

# this PR
#> # A tibble: 2 x 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 drop_empty   72.2ms   72.9ms      13.7    25.4MB     34.3
#> 2 keep_empty   81.4ms   81.4ms      12.3      29MB     86.0

# CRAN
#> Warning: Some expressions had a GC in every iteration; so filtering is disabled.
#> # A tibble: 2 x 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 drop_empty 642.71ms 642.71ms     1.56     13.1MB     21.8
#> 2 keep_empty    1.02s    1.02s     0.979    13.9MB     22.5
```

<sup>Created on 2021-06-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

### CODE REVIEW

**Comment by hadley** [comment]
How close a review do you think this needs? i.e. do you have any worries that this might change behaviour in some edge case? If so, I probably won't be able to review it until I'm next work on tidyr, since I'll need to load it all up into my brain again.

**Comment by mgirlich** [comment]
There are quite a lot of changes, so a review would definitely make sense. But it's fine for me to wait until you have time to work again on tidyr.

File: R/nest.R [hunk 1]
```diff
@@ -332,18 +332,9 @@ unnest.data.frame <- function(
...
 
-  if (nrow(data) == 0) {
-    for (col in names(cols)) {
-      data[[col]] <- as_empty_df(data[[col]], col = col)
-    }
-  } else {
-    for (col in names(cols)) {
-      data[[col]] <- map(data[[col]], as_df, col = col)
-    }
-  }
```
Comment (DavisVaughan) [inline R/nest.R:343]: There are some weird edge cases that might break if we remove this. For example, having a mix of vectors and tibbles in the list-col you are trying to unnest currently works like this:

``` r
library(tidyr)

# outer column name of `x` is used to turn `1` into `data.frame(x = 1)`
# then the unchop+unpack is done
df <- tibble(x = list(1, tibble(y = 2)))
unnest(df, x)
#> # A tibble: 2 x 2
#>       x     y
#>   <dbl> <dbl>
#> 1     1    NA
#> 2    NA     2

# again we get `data.frame(x = 1)`, which then row binds with the tibble
# since it also uses `x` as the column name
df <- tibble(x = list(1, tibble(x = 1)))
unnest(df, x)
#> # A tibble: 2 x 1
#>       x
#>   <dbl>
#> 1     1
#> 2     1

# if the column types are different, that would be an error, which leads
# me to believe that this is a "risky" feature to be relying on
df <- tibble(x = list(1, tibble(x = "x")))
unnest(df, x)
#> Error: Can't combine `..1$x` <double> and `..2$x` <character>.
```

The first two would become errors if this code was removed.

Granted, this seems like strange behavior to rely on, so maybe we can just stop supporting it? FWIW `unnest_legacy()` errors in all of these cases, which makes me think that it would be alright to not support this.

Removing this bit of code does greatly improve performance, so it would be nice to do so if possible.

CC @hadley - not sure if you can think of any other edge cases where this code mattered off the top of your head.

Comment (hadley) [reply R/nest.R:343]: I have zero recollection 😢 If removing it doesn't breaking any tests and has minimal impact on revdeps, I think it's fine to remove it. `unnest()`ing a mix of vectors and data frames doesn't seem terribly useful, and it does feel like it works by accident currently.
